### ISSUE TITLE
Add support for Optifine

### ISSUE THREAD

**Issue by JMoVS** [toolbox4minecraft/amidst#110]
I have no idea whether this issue is known or not:

```
[info] Amidst v4.0
[info] Current system time: 2016-03-01 12:24:18.473
[info] System.getProperty("os.name") == 'Mac OS X'
[info] System.getProperty("os.version") == '10.11.3'
[info] System.getProperty("os.arch") == 'x86_64'
[info] System.getProperty("java.version") == '1.8.0_73'
[info] System.getProperty("java.vendor") == 'Oracle Corporation'
[info] System.getProperty("sun.arch.data.model") == '64'
[info] Enabling OpenGL.
[info] Biome profile activated: default
[info] using '.minecraft' directory at: '/Users/justin/Library/Application Support/minecraft', libraries: '/Users/justin/Library/Application Support/minecraft/libraries'
[info] using biome profiles at: 'biome'
[info] Scanning for profiles.
[info] Successfully loaded profile list.
[info] Beginning latest version list load.
[info] Attempting to download remote version list...
[info] Successfully loaded version list. URL: https://launchermeta.mojang.com/mc/game/version_manifest.json
[info] using minecraft launcher profile 'OptiFine' with versionId '1.8.9-OptiFine_HD_U_H3'
[info] Loading libraries.
[info] Found library: /Users/justin/Library/Application Support/minecraft/libraries/optifine/OptiFine/1.8.9_HD_U_H3/OptiFine-1.8.9_HD_U_H3.jar
[info] Found library: /Users/justin/Library/Application Support/minecraft/libraries/net/minecraft/launchwrapper/1.7/launchwrapper-1.7.jar
[crash] Amidst has encounted an uncaught exception on the thread Thread[Thread-6,1,main]
[crash] java.lang.NoClassDefFoundError: io/netty/buffer/ByteBufOutputStream
    at java.lang.Class.getDeclaredFields0(Native Method)
    at java.lang.Class.privateGetDeclaredFields(Class.java:2583)
    at java.lang.Class.getDeclaredFields(Class.java:1916)
    at amidst.mojangapi.minecraftinterface.RecognisedVersion.getMainClassFields(RecognisedVersion.java:135)
    at amidst.mojangapi.minecraftinterface.RecognisedVersion.generateMagicString(RecognisedVersion.java:126)
    at amidst.mojangapi.minecraftinterface.RecognisedVersion.from(RecognisedVersion.java:120)
    at amidst.mojangapi.minecraftinterface.local.LocalMinecraftInterfaceBuilder.create(LocalMinecraftInterfaceBuilder.java:34)
    at amidst.mojangapi.file.directory.VersionDirectory.createLocalMinecraftInterface(VersionDirectory.java:106)
    at amidst.mojangapi.MojangApi.set(MojangApi.java:74)
    at amidst.gui.profileselect.LocalProfileComponent.tryLoad(LocalProfileComponent.java:82)
    at amidst.threading.WorkerExecutor.lambda$run$11(WorkerExecutor.java:69)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: io.netty.buffer.ByteBufOutputStream
    at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
    at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
    ... 14 more`
```

**Comment by BigAlanM** [comment]
My experience has been even though Optifine doesn't appear to affect the Minecraft world generator something in it interferes with Amidst and always has.
I suggest that you create a totally vanilla profile of the release of Minecraft you use and select that for Amidst to use.
That's what I've done and will continue to do.

**Comment by BigAlanM** [comment]
NOTE:  This situation WRT Optifine should probably be noted in the Wiki FAQ.

**Comment by stefandollase** [comment]
@BigAlanM is right. This is caused by Optifine. Please select an unmodded Minecraft profile. I added a note about this to [the wiki](https://github.com/toolbox4minecraft/amidst/wiki/FAQ#can-i-use-amidst-with-a-modded-minecraft-version). I close this as invalid.

**Comment by Treer** [comment]
Given the popularity of Optifine, this is one of those issues I think could be a "currently out of scope" feature request.

Since optifine doesn't change the world generation algorithms, I looked into this before 1.9 was released in there hope there might be an easy solution. 

Unfortunately this will need someone far more familiar with Java than I am, there are at least two problems...

The first problem is `NoClassDefFoundError` errors - it looks like the modified Minecraft jars expect that whatever calls them will have the following libraries available in the classpath: guava 19.0+, io.netty 4.0.34+, log4j 2.5+

Including these libraries in Amidst moves us to the next problem but makes the .exe much larger. There might be better ways to solve this problem, such as only including the classes the optifine jars reference, but IIRC the size is managable even if you just include everything (5MB I think, could be mistaken).

The next problem needs Java expertise:
```
[crash] Amidst has encounted an uncaught exception on the thread Thread[Thread-20,1,main]
[crash] java.lang.SecurityException: class "nt"'s signer information does not match signer information of other classes in the same package
        at java.lang.ClassLoader.checkCerts(Unknown Source)
        at java.lang.ClassLoader.preDefineClass(Unknown Source)
        at java.lang.ClassLoader.defineClass(Unknown Source)
        at java.security.SecureClassLoader.defineClass(Unknown Source)
        at java.net.URLClassLoader.defineClass(Unknown Source)
        at java.net.URLClassLoader.access$100(Unknown Source)
        at java.net.URLClassLoader$1.run(Unknown Source)
        at java.net.URLClassLoader$1.run(Unknown Source)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.net.URLClassLoader.findClass(Unknown Source)
        at java.lang.ClassLoader.loadClass(Unknown Source)
        at java.lang.ClassLoader.loadClass(Unknown Source)
        at java.lang.Class.getDeclaredFields0(Native Method)
        at java.lang.Class.privateGetDeclaredFields(Unknown Source)
        at java.lang.Class.getDeclaredFields(Unknown Source)
        at amidst.mojangapi.minecraftinterface.RecognisedVersion.getMainClassFields(RecognisedVersion.java:134)
        at amidst.mojangapi.minecraftinterface.RecognisedVersion.generateMagicString(RecognisedVersion.java:125)
        at amidst.mojangapi.minecraftinterface.RecognisedVersion.from(RecognisedVersion.java:119)
        at amidst.mojangapi.minecraftinterface.local.LocalMinecraftInterfaceBuilder.create(LocalMinecraftInterfaceBuilder.java:34)
        at amidst.mojangapi.file.directory.VersionDirectory.createLocalMinecraftInterface(VersionDirectory.java:106)
        at amidst.mojangapi.MojangApi.set(MojangApi.java:74)
        at amidst.gui.profileselect.LocalProfileComponent.tryLoad(LocalProfileComponent.java:82)
        at amidst.threading.WorkerExecutor.lambda$5(WorkerExecutor.java:69)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
        at java.lang.Thread.run(Unknown Source)	
```

As far as I can tell, this means that two packages are trying to access the class "nt" (which I assume is an obfuscated class name, so perhaps a Minecraft class) but they have different signatures. I have no idea how java signs things or what its security rules are.

**Comment by stefandollase** [comment]
You are right, I made it a currently out of scope feature request.

**Comment by stefandollase** [comment]
I just installed Optifine and noticed that the .json file which described the required libraries also contains this entry:

     "inheritsFrom": "1.8.9",

I guess this means we have to implement some sort of inheritance mechanism for the library loading: Use all libraries from all inherited json files. If there are conflicts, use the one which is the closest to the actually selected version. This will probably resolve the issue with the missing libraries without adding them to Amidst. However, I am not sure about the `SecurityException`. It might be resolved, but we will have to see.

**Comment by Treer** [comment]
Until Amidst supports Optifine:
* When people post their crash report, we should explain they can still use Amidst and that they'll need to create a second profile with the Launcher. Github has [Saved Replies](https://github.com/blog/2135-saved-replies) feature that can be used for template responses.
* We should trap that damn "java.lang.NoClassDefFoundError: com/google/common/util/concurrent/ListenableFuture" exception and bring up an error dialog explaining an unmodded profile needs to be created.

**Comment by stefandollase** [comment]
Thanks for the image. I adjusted my saved reply for modded Minecraft like this:

```
I close this as invalid, because Amidst does not support modded Minecraft.

You can still use Amidst with an unmodded Minecraft profile. To do so, please create a new profile with the Minecraft launcher. Call it something like "Amidst" and use that when Amidst asks you for a profile. Make sure to select the correct Minecraft version when creating the new profile.

![minecraft launcher](https://cloud.githubusercontent.com/assets/6390507/14304860/0cd746e8-fbfa-11e5-8451-e5f97cdeeb03.png)

For more information, please read the FAQ: [Can I use Amidst with a modded Minecraft version?](https://github.com/toolbox4minecraft/amidst/wiki/FAQ#can-i-use-amidst-with-a-modded-minecraft-version)
```

It looks like this:

>I close this as invalid, because Amidst does not support modded Minecraft.
>
>You can still use Amidst with an unmodded Minecraft profile. To do so, please create a new profile with the Minecraft launcher. Call it something like "Amidst" and use that when Amidst asks you for a profile. Make sure to select the correct Minecraft version when creating the new profile.
>
>![minecraft launcher](https://cloud.githubusercontent.com/assets/6390507/14304860/0cd746e8-fbfa-11e5-8451-e5f97cdeeb03.png)
>
>For more information, please read the FAQ: [Can I use Amidst with a modded Minecraft version?](https://github.com/toolbox4minecraft/amidst/wiki/FAQ#can-i-use-amidst-with-a-modded-minecraft-version)

---

Displaying a more helpful error message for the modded Minecraft crashes is also already on my list :-P

**Comment by PizzaCrust** [comment]
@stefandollase @Treer 
make them remove the META-INF, META-INF isn't used anymore and contains security certificates that disallows any modification of the JAR and causes a `SecurityException`.

**Comment by stefandollase** [comment]
@PizzaCrust
Can you clarify? Do you want us to ask Mojang to remove the META-INF, because they don't use it anymore? Or do you want us to modify the Minecraft jar file when it is used by Amidst? In the later case: You might be right about this. Unfortunately, this requires us to copy the jar file, before using it, because we should not modify the original jar file. This also implies, we need a location in the file system to store the copied jar file :-/

I still have no experience or knowledge about code signing in Java. When I get the time, I will read and learn about it.

**Comment by PizzaCrust** [comment]
@stefandollase 
mods before **Minecraft Forge** had to be manually implemented into the JAR and had the META-INF deleted because of the certificates. META-INF's Manifest isn't required since the launcher uses JSON files to launch Minecraft. I suggest that you guys create a dialog that informs the user that the target JAR version will be modified and do some nasty copying, extracting, and creating a new JAR.

**Comment by stefandollase** [comment]
@PizzaCrust
We might be able to just load the inherited Minecraft version: The user selects his Optifine profile, but we directly load the unmodded Minecraft version. We don't need to load the modded version, because Optifine does not alter the biome generator.

If this works as described, we don't need to copy and alter the jar file, which I would prefer.

Anyway, thanks for the pointer! I just remembered that I manually modded Minecraft quite some time ago and I also had to remove the META-INF directory. I completely forgot about this :-P

**Comment by stefandollase** [comment]
This was implemented in #332 and is part of the latest release [Amidst v4.3-beta2](https://github.com/toolbox4minecraft/amidst/releases/tag/v4.3-beta2).

I also updated the wiki to reflect the current state for modded Minecraft:

>## [Why can I not use Amidst with a modded Minecraft version?](https://github.com/toolbox4minecraft/amidst/wiki/Using-Amidst#why-can-i-not-use-amidst-with-a-modded-minecraft-version)
>
>In general, Amidst is not designed to support modded Minecraft versions. However, when a modded Minecraft profile is selected, Amidst will try to load the underlying vanilla Minecraft version instead. This is especially useful for a mod like Optifine, which does not alter the biome generation. However, if the modded Minecraft profile uses a different biome generator than the corresponding vanilla Minecraft version, Amidst will still show you the vanilla map, which is different from the actually generated world.

I only tested this with an Optifine profile. It would be great to receive feedback about whether it also works with other modded Minecraft profiles :-) (preferably in a new issue)
